Fix gui alignment issues

feerrenrut opened the pull request:
Fixed alignment and made small modifications to several nvda dialogs.

Also fixes:
#6317 (Gestures treeview does not look like a treeview)
#5548 (Text control to specify the path when creating a portable copy is quite narrow)
#6342 (Spacing between labels and combo boxes)
#6343 (Inconsistency with vertical spacing between checkbox/label options)
#6349 (input gestures dialog tree view too small)

feerrenrut committed:
fix alignment of the exit dialog

feerrenrut committed:
fix general settings alignments

feerrenrut committed:
fixed alignment in the synthesizer dialog

feerrenrut committed:
fix alignment on voice settings dialog

feerrenrut committed:
fix alignment on keyboard and mouse settings dialogs

feerrenrut committed:
fix alignment on braille settings dialogs

feerrenrut committed:
fix alignment on the object presentation settings dialogs

feerrenrut committed:
fix alignment on the dictionary dialog

feerrenrut committed:
fix alignment on the input gestures dialog

feerrenrut committed:
fix alignment on the configuration profiles dialog

feerrenrut committed:
fix alignment on the new config profile dialog

feerrenrut committed:
fix alignment on the profile triggers dialog

feerrenrut commented:
Images of the changed dialogs: 
![profile triggers](https://cloud.githubusercontent.com/assets/3491346/17848142/7992e0b6-6884-11e6-97fb-7d6230f5d91d.png)
![synthesizer settings](https://cloud.githubusercontent.com/assets/3491346/17848143/7997d9e0-6884-11e6-8c88-0579efb60f7c.png)
![voice settings](https://cloud.githubusercontent.com/assets/3491346/17848146/79a4ab66-6884-11e6-9e91-12e2d954f47a.png)
![braille settings](https://cloud.githubusercontent.com/assets/3491346/17848144/799a23b2-6884-11e6-9b21-865ce493eb8c.png)
![configurations](https://cloud.githubusercontent.com/assets/3491346/17848147/79a67db0-6884-11e6-8846-170977eb4f6a.png)
![dictionary entries](https://cloud.githubusercontent.com/assets/3491346/17848145/79a1605a-6884-11e6-9082-94b96b6a1dc9.png)
![exit](https://cloud.githubusercontent.com/assets/3491346/17848148/79c69352-6884-11e6-98be-0bae139495bb.png)
![general settings](https://cloud.githubusercontent.com/assets/3491346/17848150/79d13334-6884-11e6-85c7-a35f8b6882e7.png)
![input gestures](https://cloud.githubusercontent.com/assets/3491346/17848149/79cfb518-6884-11e6-9d13-5bedee8a8cee.png)
![keyboard settings](https://cloud.githubusercontent.com/assets/3491346/17848151/79d51e68-6884-11e6-9479-1f74f3ddce3b.png)
![mouse settings](https://cloud.githubusercontent.com/assets/3491346/17848153/79dd5344-6884-11e6-80a2-5edd306090a2.png)
![new profile](https://cloud.githubusercontent.com/assets/3491346/17848152/79db3050-6884-11e6-82a7-b3e8ccc4743b.png)
![object presentation](https://cloud.githubusercontent.com/assets/3491346/17848154/79fc3674-6884-11e6-95ab-b2572d31ced7.png)

jcsteh commented:
Thanks for the fixes! :) First, a few specific concerns:

1. Configuration Profiles dialog: I see you've "grouped" the list and the buttons for manipulating profiles using a StaticBoxSizer. I sort of follow this, since Triggers are separate from that list. The problem is that this gets reported as a grouping for screen reader users, which is IMO extraneous verbosity here. Groupings certainly make sense when it needs to be made clear that a set of options are in a group, but in this case, everything is really related to the same thing: profiles. Perhaps it might be helpful if you can explain how this "looks better" visually. Does it make something clear that wasn't clear before?
2. Input Gestures dialog: All of the controls except the OK and Cancel buttons are now inside a "Gestures" box (exposed to screen reader users as a grouping). Again, can you explain how this "looks" better visually? There's nothing else in the dialog, so I'm just trying to follow the need to group things here. Again, my concern is that there is some extraneous verbosity here for screen reader users.
3. dropDownLabelBorder and dropDownLabelFlags get defined in quite a few places. If these are the border/flags we need to use for drop-downs everywhere, I think these should be moved into the SettingsDialog class as class constants. That way, they can be reused.

Beyond this (and perhaps something that needs to be addressed separately), a major concern for me here is that as a blind developer, I'm struggling to understand the "rules" we need to use to make the GUI look decent. Obviously, we're always going to need some help to "pretty things up", but it'd be good if we can avoid making the same mistakes over and over. Perhaps we need to develop a set of guidelines as to what borders/spacing/flags/proportions to use and when to use them, etc. As another example, I don't follow when we should use a border and when we should add a "spacer". (I didn't even know about spaces before today! :)) I wonder how much of this we can somehow hide behind code. Perhaps we could create our own helper methods or sizer subclasses or the like.

Thoughts?

nishimotz commented:
Regarding dropDownLabelBorder and dropDownLabelFlags:
When wx.StaticText and wx.Choice are used horizontally together like those cases, this work makes it nicer visually.
In such cases, the static text should be used as the label of the following combobox.
I am in favor of refactoring SettingDialog class regarding this combination, because it is semantically meaningful.

feerrenrut committed:
Pull out common gui patterns into a helper module.

feerrenrut commented:
@jcsteh As per our conversation I have pulled the common patterns out into a helper, and removed one of the group boxes. The other I have left in place (since it helps to visually associate the items) but removed the label. The configProfiles code is harder to make generic.

Could you please take another look when you get the chance?

feerrenrut committed:
Show symbols on tree view

To resolve visual ambiguity between a listview and a treeview, show plus/minus symbols

feerrenrut committed:
fixed up alignment and spacing on the elements list dialog

feerrenrut commented:
See #6342

feerrenrut commented:
See #6343 (Inconsistency with vertical spacing between checkbox/label options)

feerrenrut commented:
See #6349 (input gestures dialog tree view too small)

feerrenrut commented:
Feedback on GUI, received via email, with respect to the create portable nvda dialog
> this dialog seems to have no real padding around any of the items and the edges of the dialog. we need to make sure we have a consistent padding around the items in the dialog like with all other major dialogs. The issue of spacing exists here too in regards to the path edit field and the browse button.  there needs to be more space between them like in the combo-box / text label issue already discussed.  the continue/cancel also needs to be padded out (similar to the ok/cancel issue already discussed) and again, right aligned.
> 
> Also the text along the top does wrap in the dialog making the layout seem messy. maybe a re-wording would solve this. 
> e.g.: 
> Select a folder for the portable copy of NVDA 
> [then you have under this the edit field and browse button] 
> Use the current configuration for the portable copy (which is a label with a checkbox to its left as per normal) 
> Continue/Cancel (right aligned)

- [ ] Fix up the portable nvda dialog
- [ ] Look for other dialogs that may have been missed

feerrenrut committed:
WIP

feerrenrut committed:
Gui helper is now class based

feerrenrut committed:
Moved more of the code over to the class based approach

feerrenrut committed:
Updated comments for the guiHelper

michaelDCurran commented on source/gui/settingsDialogs.py:
```diff
@@ -129,80 +133,82 @@ class GeneralSettingsDialog(SettingsDialog):
 		# Translators: The list of languages for NVDA.
-		self.languageList=wx.Choice(self,languageListID,name=_("Language"),choices=[x[1] for x in self.languageNames])
+		languageCtrlName = _("Language")
+		self.languageList=settingsSizerHelper.addLabeledControl(languageLabelText, wx.Choice, name=languageCtrlName, choices=languageChoices)
```
the name=languageControlName for wx.Choice is not needed (assuming it is never shown visually).

michaelDCurran commented on source/gui/settingsDialogs.py:
```diff
@@ -129,80 +133,82 @@ class GeneralSettingsDialog(SettingsDialog):
-		self.logLevelList=wx.Choice(self,logLevelListID,name=_("Log level"),choices=[name for level, name in self.LOG_LEVELS])
+		logLevelChoicesName=_("Log level")
+		logLevelChoices = [name for level, name in self.LOG_LEVELS]
+		self.logLevelList = settingsSizerHelper.addLabeledControl(logLevelLabelText, wx.Choice, name=logLevelChoicesName, choices=logLevelChoices)
```
name is not needed.

michaelDCurran commented on source/gui/settingsDialogs.py:
```diff
@@ -451,12 +447,13 @@ def makeSettingControl(self,setting):
 		setattr(self,"_%ss"%setting.name,getattr(synth,"available%ss"%setting.name.capitalize()).values())
 		l=getattr(self,"_%ss"%setting.name)###
-		lCombo=wx.Choice(self,wx.ID_ANY,name="%s:"%setting.i18nName,choices=[x.name for x in l])
+		labeledControl=guiHelper.LabeledControlHelper(self, labelText, wx.Choice, name="%s:"%setting.i18nName, choices=[x.name for x in l])
```
name can be removed

michaelDCurran commented on source/gui/settingsDialogs.py:
```diff
@@ -618,85 +618,95 @@ class KeyboardSettingsDialog(SettingsDialog):
-		# Translators: This is the name of a combobox in the keyboard settings dialog.
-		self.kbdList=wx.Choice(self,kbdListID,name=_("Keyboard layout"),choices=[layouts[layout] for layout in self.kbdNames])
+		kbdChoices = [layouts[layout] for layout in self.kbdNames]
+		self.kbdList=sHelper.addLabeledControl(kbdLabelText, wx.Choice, name=kbdChoicesName, choices=kbdChoices)
```
name can be removed

michaelDCurran commented on source/gui/settingsDialogs.py:
```diff
@@ -732,47 +742,52 @@ class MouseSettingsDialog(SettingsDialog):
-		# Translators: This is the name of a combobox in the mouse settings dialog.
-		self.textUnitComboBox=wx.Choice(self,wx.ID_ANY,name=_("text reporting unit"),choices=[textInfos.unitLabels[x] for x in self.textUnits])
+		textUnitsChoices = [textInfos.unitLabels[x] for x in self.textUnits]
+		self.textUnitComboBox=sHelper.addLabeledControl(textUnitLabelText, wx.Choice, name=textUnitsName, choices=textUnitsChoices)
```
name can be removed

michaelDCurran commented on source/gui/settingsDialogs.py:
```diff
@@ -888,63 +903,69 @@ class ObjectPresentationDialog(SettingsDialog):
-		self.progressList=wx.Choice(self,progressListID,name=_("Progress bar output"),choices=[name for setting, name in self.progressLabels])
+		progressName=_("Progress bar output")
+		progressChoices = [name for setting, name in self.progressLabels]
+		self.progressList=sHelper.addLabeledControl(progressLabelText, wx.Choice,name=progressName,choices=progressChoices)
```
name can be removed

feerrenrut committed:
Fix some errors and remove 'name' parameter

Some wx controls were being created using the 'name' parameter. When
there is a label this is unnecessary, since the label will describe the
controls. It may also be unnecessary when there is sufficient context
provided by the dialog itself.

feerrenrut committed:
Fixed up missed part of gui

Label for addons disabled was missed in clean up. This was not easy to
notice since it is conditionally shown.

feerrenrut committed:
fix alignment on several dialogs

Alignments fixed on symbol/pronunciation dialog.
Fixed alignment on 'make portable copy' dialog

Groups of controls (wx.StaticBoxSizer) added to a VERTICAL boxSizerHelper are now
automatically expanded to the full width. I think this is a more common
pattern.

Directory entry concept has been abstracted.

feerrenrut committed:
Update some remainig dialogs

Fixed up the:
- launcher
- installer dialog
- update check result

feerrenrut committed:
minor fixes

Updates to the user guide, and a the symbol/pronunciation add symbol
dialog.

feerrenrut commented:
Consider also closing: #5548

jcsteh commented:
nit: There are quite a few added blank lines (particularly in `gui/settingsDialogs.py`) that have tabs on them. I'm not too worried about this, but if there's an easy way for you to remove the tabs, please do. A regular expression like ^\w+$ should catch this.

jcsteh commented on source/gui/settingsDialogs.py:
```diff
@@ -1417,130 +1427,106 @@ class BrailleSettingsDialog(SettingsDialog):
-		settingsSizer.Add(item, border=10, flag=wx.BOTTOM)
+		readByParagraphText = _("Read by &paragraph")
+		self.readByParagraphCheckBox = sHelper.addItem(wx.CheckBox(self, label=readByParagraphText))
+		self.readByParagraphCheckBox.Value = config.conf["braille"]["readByParagraph"]
```
I'm totally cool with just accepting these as stylistic differences, but I'm curious nevertheless:

1. You've changed most places to assign label text to a variable, rather than just specifying the label in the call. Why? I guess this makes the lines shorter, but you could wrap to a new line. This approach seems more "indirect" to me. It also creates a lot more local variables, but that's inconsequential; premature optimisation and all. :)
2. Here (and in quite a few other places), I had a variable like `item` which I just assign temporarily. The reason is largely to make the code less verbose and thus easier to type. You've changed these to either use a new local or, if there was an attribute on `self`, to use that. Why? Was the use of `item` confusing to you; i.e. did you feel you weren't sure which control I was referring to despite the grouping of the code? One other reason I did this was to avoid hash lookups on `self`, but again, inconsequential; premature optimisation and all. :)

Again, not asking for changes here; just curious on your reasoning.

feerrenrut replied:
For 1: yes, to make the lines shorter. But also to try to better pair the text with the translator comment.

For 2: Using the named variable rather than `item` makes the code easier to read and understand. When `item` is used, and then a few lines down something is being done to it, you have to remember (or go look for) the context. With the named variable, (hopefully) it is fairly self explanatory what you are dealing with.

jcsteh reviewed (changes requested):
Looks great overall! Thanks!

feerrenrut commented on source/gui/configProfiles.py:
```diff
@@ -367,10 +374,11 @@ def __init__(self, parent):
 		item = self.profileName = wx.TextCtrl(self)
 		sizer.Add(item)
-		mainSizer.Add(sizer)
+		mainSizer.Add(sizer, border=10, flag=wx.ALL)
```
It looks like I missed this one when converting to the guiHelper. I have updated this.

feerrenrut committed:
Review actions for #6287

jcsteh commented on source/gui/guiHelper.py:
```diff
@@ -1,18 +1,17 @@
-import wx
-
 """ Example usage
+Utilities to simplify the creation of wx GUIs, including automatic management of spacing.
```
These two lines should probably be reversed; i.e. so "Example usage" comes below the summary.

jcsteh reviewed (approved):
Looks great!

feerrenrut committed:
Fixed up an out of order comment for #6287

feerrenrut commented:
This has been merged into master with commit: 87ed4d17feee64b4ea021a125baed1dbba18a187
